flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerReleaseInSlotManagerTest.java

```diff
@@ -0,0 +1,165 @@
+			executor.execute(() -> slotManager.registerTaskManager(taskManagerConnection, slotReport));
+			assertThat(releaseFuture.get(), is(equalTo(taskManagerConnection.getInstanceID())));
+		}
+	}
+
+	/**
+	 * Tests that idle but not releasable task managers will not be released even if timed out before it can be.
+	 */
+	@Test
+	public void testTaskManagerIsNotReleasedBeforeItCanBe() throws Exception {
+		try (SlotManager slotManager = createAndStartSlotManagerWithTM()) {
+			testTaskManagerIsNotReleasedBeforeItCanBe(slotManager);
+			testTaskManagerReleasedAfterItCanBe(slotManager);
+		}
+	}
+
+	/**
+	 * Tests that idle task managers will not be released after "can be" check in case of concurrent resource allocations.
+	 */
+	@Test
+	public void testTaskManagerIsNotReleasedInCaseOfConcurrentAllocation() throws Exception {
+		try (SlotManager slotManager = createAndStartSlotManagerWithTM()) {
+			resetCanBeReleasedFutureAndTriggerTimeoutCheck(slotManager);
+
+			// Allocate and free slot between triggering TM.canBeReleased request and receiving response.
+			// There can be potentially newly unreleased partitions, therefore TM can not be released yet.
+			AllocationID allocationID = new AllocationID();
+			slotManager.registerSlotRequest(new SlotRequest(new JobID(), allocationID, resourceProfile, "foobar"));
+			mainThreadExecutor.triggerAll();
+			slotManager.freeSlot(slotId, allocationID);
+
+			setCanBeReleasedAndTrggerCheckTmReleased(true, false);
+
+			testTaskManagerReleasedAfterItCanBe(slotManager);
+		}
+	}
+
+	private SlotManager createAndStartSlotManagerWithTM() {
+		SlotManager slotManager = SlotManagerBuilder
+			.newBuilder()
+			.setScheduledExecutor(mainThreadExecutor)
+			.setTaskManagerTimeout(Time.milliseconds(0L))
+			.build();
+		slotManager.start(resourceManagerId, mainThreadExecutor, resourceManagerActions);
+		mainThreadExecutor.execute(() -> slotManager.registerTaskManager(taskManagerConnection, slotReport));
+		return slotManager;
+	}
+
+	private void resetCanBeReleasedFutureAndTriggerTimeoutCheck(SlotManager slotManager) {
+		canBeReleased.set(new CompletableFuture<>());
+		mainThreadExecutor.execute(slotManager::checkTaskManagerTimeouts); // trigger TM.canBeReleased request
+		mainThreadExecutor.triggerAll();
+	}
+
+	private void testTaskManagerIsNotReleasedBeforeItCanBe(SlotManager slotManager) {
+		resetCanBeReleasedFutureAndTriggerTimeoutCheck(slotManager);
+		setCanBeReleasedAndTrggerCheckTmReleased(false, false);
+	}
+
+	private void testTaskManagerReleasedAfterItCanBe(SlotManager slotManager) {
+		resetCanBeReleasedFutureAndTriggerTimeoutCheck(slotManager);
+		setCanBeReleasedAndTrggerCheckTmReleased(true, true);
+	}
+
+	private void setCanBeReleasedAndTrggerCheckTmReleased(boolean canBeReleasedToSet, boolean tmReleased) {
+		canBeReleased.get().complete(canBeReleasedToSet);
+		mainThreadExecutor.triggerAll();
+		assertThat(releaseFuture.isDone(), is(tmReleased));
```

Comment:
I would suggest to separate "acting" as well as "asserting" from each other. The test case is simpler to understand if the assertions are done after we have completed the "acting" in the test cases. E.g.

```
// acting
setTaskManagerCanBeReleasedAndTriggerTimeoutCheck();

// asserting
verifyTaskManagerHasBeenReleased();
```